tools/bss_reordering.py

```diff
@@ -0,0 +1,191 @@
+    with open(object_path, "rb") as f:
+        elffile = elftools.elf.elffile.ELFFile(f)
+        symtab = elffile.get_section_by_name(".symtab")
+        data = elffile.get_section_by_name(section_name).data()
+
+        reloc_section = elffile.get_section_by_name(f".rel{section_name}")
+        if reloc_section is None:
+            return []
+
+        relocs = []
+        offset_hi16 = 0
+        for reloc in reloc_section.iter_relocations():
+            reloc_offset = reloc.entry["r_offset"]
+            reloc_type = reloc.entry["r_info_type"]
+            reloc_name = symtab.get_symbol(reloc.entry["r_info_sym"]).name
+
+            if reloc_type == 2:  # R_MIPS_32
+                offset_32 = reloc_offset
+                addend = int.from_bytes(
+                    data[reloc_offset : reloc_offset + 4], "big", signed=True
+                )
+                relocs.append(Reloc(reloc_name, offset_32, None, None, addend))
+            elif reloc_type == 5:  # R_MIPS_HI16
+                offset_hi16 = reloc_offset
+            elif reloc_type == 6:  # R_MIPS_LO16
+                address_lo16 = reloc_offset
+                addend_hi16 = int.from_bytes(
+                    data[offset_hi16 + 2 : offset_hi16 + 4], "big", signed=False
+                )
+                addend_lo16 = int.from_bytes(
+                    data[address_lo16 + 2 : address_lo16 + 4], "big", signed=True
+                )
+                addend = (addend_hi16 << 16) + addend_lo16
+                relocs.append(
+                    Reloc(reloc_name, None, offset_hi16, address_lo16, addend)
+                )
+
+        return relocs
+
+
+def read_u32(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(4), "big")
+
+
+def read_u16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big")
+
+
+def read_s16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big", signed=True)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Report BSS reorderings between the baserom and the current build "
+        "by parsing relocations from the built object files and comparing their final values "
+        "between the baserom and the current build. "
+        "Assumes that the only differences are due to BSS ordering."
+    )
+    parser.add_argument(
+        "--oot-version",
+        "-v",
+        type=str,
+        default="gc-eu-mq-dbg",
+        help="OOT version (default: gc-eu-mq-dbg)",
+    )
+    parser.add_argument(
+        "--segment",
+        type=str,
+        help="ROM segment to check, e.g. 'boot', 'code', or 'ovl_player_actor' (default: all)",
+    )
+
+    args = parser.parse_args()
+    version = args.oot_version
+
+    mapfile = mapfile_parser.mapfile.MapFile()
+    mapfile.readMapFile(f"build/{version}/oot-{version}.map")
+
+    segments = []
+    for segment in mapfile:
```

Comment:
As a note and this is a problem in mapfile-parser's terminology, these "segments" corresponding to rom segments are elf sections, elf segments (also known as program headers) are something else

Reply:
Should I rename these variables or is it fine for now?

Reply:
I changed this to `mapfile_segment`, since it also helps distinguish it from `args.segment` (which is a "ROM segment" or "spec segment" so I think that name is correct)